### ISSUE TITLE
More information on MFX_conf_t variables (MotionFX v1.0.1 )

### ISSUE THREAD

**Issue by mmrobinson-gi** [STMicroelectronics/ispu-examples#1]
Hello ST,

We've been using the MotionFX library for orientation estimation (yaw, pitch, and roll) on an ISM330IS - with only an accelerometer and gyroscope.

The introduction of MotionFX v1.0.1 introduced some new factors to play around with...

Snippet from https://github.com/STMicroelectronics/ispu-examples/blob/master/ism330is_lsm6dso16is/sensor_fusion_6x/ispu/lib/motion_fx.h

> typedef struct
> {
>   float alpha_acc;              /* Coefficient to control accelerometer mean and variance computation [0 to 1] */
>  float alpha_gyr;              /* Coefficient to control gyroscope mean and variance computation [0 to 1] */
>   float alpha_mag;              /* Coefficient to control magnetometer mean and variance computation [0 to 1] */
>   float kappa_acc;              /* Coefficient to control accelerometer weighting factor [0 to 1] */
>   float kappa_mag;              /* Coefficient to control magnetometer weighting factor [0 to 1] */
>   float acc_max;                /* Linear acceleration threshold [g] */
>   float acc_var_max;            /* Linear acceleration var-based threshold [g] */
>   float mag_var_max;            /* Magnetic anomaly var-based threshold [g] */
>   float gb_alpha;               /* Gbias calibration low-pass filter [0 to 1] */
>   float gb_acc_max;             /* Gbias calibration acceleration norm-based threshold [g] */
>   float gb_gyr_max;             /* Gbias calibration angular rate threshold [dps] */
>   float gb_acc_var_max;         /* Gbias calibration acceleration var-based threshold [g] */
>   float gb_gyr_var_max;         /* Gbias calibration angular rate var-based threshold [dps] */
>   float gb_gyr_var_max_greedy;  /* Gbias calibration initial angular rate var-based threshold [dps] */
>   float gb_time;                /* Gbias calibration time-based threshold [s] */
> } MFX_conf_t;

Would it be possible to share a brief description of what each of the variables in the MFX_conf_t tend to affect in the performance of the sensor fusion algorithm for orientation estimation?

Kind regards,

Matt

**Comment by mmrobinson-gi** [comment]
Also would like to add another question - Is there any conflict between setting the gbias configurations here and enabling the gbias calibration at the same time?

**Comment by lorenzobracco** [comment]
Hello Matt,

Here are some additional details about the parameters:
- alpha_acc: the smaller the value, the slower the filter is for computing mean and variance, which are used for determining against acc_max and acc_var_max thresholds if there are linear accelerations, and thus whether accelerometer is used to update the orientation estimation. Also, they are used for stationary detection for gbias estimation.
- alpha_gyr: analogous to alpha_acc. Mean and variance are used for stationary detection for gbias estimation.
- alpha_mag: analogous to the other alpha parameters, but for the norm of magnetometer data. The computed variance is then used to determine if a magnetic anomaly is detected (comparison with mag_var_max), and thus whether magnetometer is used to update the orientation estimation.
- kappa_acc: when linear accelerations are not detected, it determines the weight of the accelerometer in the fusion filter. A lower value will make the library use accelerometer less and rely more on gyroscope data.
- kappa_mag: when magnetic anomalies are not detected, it determines the weight of the magnetometer in the fusion filter. A lower value will make the library use magnetometer less and rely more on gyroscope data.
- acc_max: threshold on accelerometer norm for detecting linear accelerations (also see alpha_acc explanation). A lower value will cause accelerometer to be used less (the device must be more stationary).
- acc_var_max: threshold on accelerometer variance for detecting linear accelerations (also see alpha_acc explanation). A lower value will cause accelerometer to be used less (the device must be more stationary).
- mag_var_max: threshold on magnetometer norm variance for detecting magnetic anomalies (also see alpha_mag explanation). A lower value will make the library more sensitive to anomalies, with the increased risk of false positives.
- gb_alpha: controls the low-pass filter on the gbias estimation. A lower value makes the estimation more stable, but less responsive to changes.
- gb_acc_max: threshold on accelerometer norm for detecting stationary condition to estimate gbias. A lower value means that the device must be more stationary for the gbias estimator to work.
- gb_gyr_max: threshold on gyroscope for detecting stationary condition to estimate gbias. A lower value means that the device must be more stationary for the gbias estimator to work.
- gb_acc_var_max: threshold on accelerometer variance for detecting stationary condition to estimate gbias. A lower value means that the device must be more stationary for the gbias estimator to work.
- gb_gyr_var_max: threshold on gyroscope variance for detecting stationary condition to estimate gbias. A lower value means that the device must be more stationary for the gbias estimator to work.
- gb_gyr_var_max_greedy: threshold on gyroscope variance for detecting stationary condition for the first time to estimate the gbias. A lower value means that the device must be more stationary for the gbias estimator to work.
- gb_time: time needed below the thresholds above to detect stationary condition to estimate gbias.

To sum up, you mainly have control over how much accelerometer is used in the fusion (how much stationary the device needs to be to use it and how much it is weighted when it is stationary) and over how much stationary the device needs to be for gbias estimation. The relation between accelerometer usage and final performance is very dependent on the use case. In terms of gbias configuration, you want to be able to estimate gbias when very still, but not when moving to avoid a bad estimation.

There is no conflict between this configuration and the enabling of the gbias. The configuration parameters do configure the gbias estimator that you enable via the dedicated API (MotionFX_enable_gbias).

I hope this answers your questions.

Thanks,
Lorenzo

**Comment by mmrobinson-gi** [comment]
Thank you so much Lorenzo. Very helpful and should help the optimisation process.
